### PR TITLE
Add 'lexer' and 'registered' DSL keywords

### PR DESCRIPTION
**Pull request by pyrmont**
As discussed in #1135 and #1147, the large number of lexers provided by Rouge causes the loading of the library to use a substantial amount of memory. This is particularly wasteful given that most uses of Rouge will only be to highlight a handful of languages.

This PR proposes the addition of two new keywords to the DSLs used in Rouge to help alleviate this problem:

- The first keyword is the **`lexer` keyword**. It forms part of the generic lexer DSL and is intended to be used to wrap the lexing logic of a given lexer. By putting the logic inside a block, we defer the instantiation of objects defined in that block until an instance of the lexer is actually created.

   An abstract implementation is defined in `Rouge::Lexer` and a concrete implementation is defined in `Rouge::RegexLexer`. The basic idea is that the logic in the block is saved into an instance variable on the lexer's class itself. When a lexer of this class is initialised, the block is called if present. After being called the first time, the block is no longer required and is removed (to avoid repeated calls for future instantiations).

- The second keyword is the **`registered` keyword**. It forms part of the regular expression lexer DSL and is intended to be used to wrap registered words used by a given lexer. By putting these strings inside a block, we defer their instantiation until the registered words are required. 

   The implementation is defined in `Rouge::RegexLexer`. The basic idea is to create an instance method on the lexer's class that memoises the value of the registered words. This is equivalent to the following pattern:

   ```rb
   def self.keywords
     @keywords ||= %w(some example keywords)
   end
   ```

   One of the advantages of using `registered` is that the method created is available to instances without needing to preface calls with `self.class`.

The use of both the `lexer` keyword and the `registered` keyword is optional and does not break existing lexers. This will allow the progressive enhancement of lexers in Rouge while being immediately available for use by new lexer authors.

An example of how they could be used is below:

```rb
module Rouge
  module Lexers
    class WrappedLexer < RegexLexer
      title 'Wrapped'
      tag 'wrapped'

      lexer do
        registered :keywords do
          %w(some example keywords)
        end

        state :root do
          rule /\s+/, Text

          rule /(\w+)/ do |m|
            if keywords.include? m[0]
              token Keyword
            else
              token Text
            end
          end
        end
      end
    end
  end
end
```

### CODE REVIEW

**Comment by pyrmont** [comment]
What do you think @jneen and @ashmaroli? :)

**Comment by jneen** [comment]
I really like the idea - will find time to review in more detail soon.

**Comment by jneen** [comment]
Can you confirm that a string literal inside of a block takes up less memory than a string literal in an immediate context? I would imagine the block would need to store approximately the same amount of data.

**Comment by ashmaroli** [comment]
@pyrmont I like the idea as well.
However, not really sure about the impact of *`logic in the block is saved into an instance variable on the lexer's class itself`* on **retained memory usage** since objects stashed won't be GC'ed.

From the first pass, I'm confident that there shouldn't be much to worry about if caching occurs only for those lexers that are invoked by the user.

:+1: :+1:

**Comment by pyrmont** [comment]
@jneen From my initial (and admittedly limited) testing, I believe that a string literal inside a block takes up less space if you discount the overhead of the block. 

In this code:

```rb
require 'memory_profiler'
  
report = MemoryProfiler.report do
  class A
    def self.wrap(&blk)
      @wrapped = blk
    end
   
    wrap do
      %w(
        Here are some strings that take up space
      )
    end
  end
   
  class B
    strings = %w(
      Here are some strings that take up space
    )
  end
end
         
report.pretty_print(scale_bytes: true)
```

the memory usage of A doesn't increase regardless of how many words are in the block passed to `wrap` in A. In contrast, adding new words to `strings` in B does increase usage.

I've tried to find information online about the memory usage of objects inside blocks but without much success.

**Comment by pyrmont** [comment]
@ashmaroli The lexing logic inside the block passed to `lexer` by each block is attached as an instance variable and so there is the overhead of the block itself.

As mentioned in the previous comment, I've struggled to find much information about the memory allocation of objects _inside_ blocks _before_ the block is called. My limited testing appeared to show that the objects are _not_ created. My guess is that they remain in Ruby's AST and aren't actually created until the block is called.

If and when the block is called, it's then removed from the class. Although at this point the objects created by the block have now been added to the class so it should be 'bigger' at this point than it was before.

**Comment by jneen** [comment]
Perhaps it's a difference between checking the memory usage *of the class* vs the total memory usage involved. For example, I'm curious as to the difference between these two

``` ruby
class A
  @foo = lambda { %w(... very large list ...) }
end

class B
  @foo = C.new %w(... very large list ...)
end
```

**Comment by jneen** [comment]
I understand that object overhead is a big thing - but if the string data is very large (as is the case in many lexers) I worry that it would be about the same.

**Comment by pyrmont** [comment]
Alas, looks like there may be no such thing as a free lunch.

I took the easiest to modify lexers (122 files), used a script to insert a `lexer do… end` block around the lexing logic and then compared the memory usage when loading up a version of the library containing only these lexers.

Memory goes from 9.56MB to 8.94MB :(

**Comment by ashmaroli** [comment]
> Memory goes from 9.56MB to 8.94MB :(

@pyrmont Can you open a PR with the updated library? So that I can manually compare the memprof report from the Travis build..?

**Comment by ashmaroli** [comment]
> The lexing logic inside the block passed to `lexer` by each block is attached as an instance variable and so there is the overhead of the block itself.

So what would that entail for the following code? Would the memory usage remain the same or grow..?:
```ruby
(1..10).to_a.each do |i|
  Rouge::Lexers::Ruby.new.lex(code_library[i])
end
```

**Comment by pyrmont** [comment]
Here you go:

- [lexers2.tar.gz](https://github.com/rouge-ruby/rouge/files/3246182/lexers2.tar.gz)
- [lexers3.tar.gz](https://github.com/rouge-ruby/rouge/files/3246189/lexers3.tar.gz)

Files will expand to `lib/rouge/lexers2` and `lib/rouge/lexers3` respectively. You need to change the loading directory in `lib/rouge.rb` and `lib/rouge/lexer.rb`.

**Comment by pyrmont** [comment]
> So what would that entail for the following code? Would the memory usage remain the same or grow..?:
> 
> ```rb
> (1..10).to_a.each do |i|
>   Rouge::Lexers::Ruby.new.lex(code_library[i])
> end
> ```

Well, when `Rouge::Lexers::Ruby` is loaded, the block passed to `lexer` is assigned to an instance variable called `@lexer` on the `Rouge::Lexers::Ruby` singleton class. When `.new` is called for the first time on `Rouge::Lexers::Ruby`, the `Rouge::Lexer#initialize`
method is called and the initialiser gets the block from `@lexer`, calls it and then removes `@lexer`.

Calling the block mostly invokes calls to `RegexLexer.state`. The net result of these calls is to add the blocks passed to `state` to the state definitions instance variable on the `Rouge::Lexers::Ruby` singleton class. These definitions are used by the instances of `Rouge::Lexers::Ruby` as is sort of visible here:

https://github.com/rouge-ruby/rouge/blob/8800d7f3bdc15c81f633d6eba93641f07b7b40bc/lib/rouge/regex_lexer.rb#L202-L222

Basically, Rouge's DSL works by adding state definitions to the lexer's singleton class. My proposed ‘lexer` method (and the changes to `Lexer#initialize` produce the same result but only _after_ the system tries to create an instance of a lexer. the system will only try to do this when the consumer tries to highlight source text that will use the lexer and so this avoids adding objects to the lexer singleton class until we know we need them.

**Comment by pyrmont** [comment]
Sorry, I revised the previous comment quite a bit. Hopefully it's clearer now.

**Comment by ashmaroli** [comment]
> to an instance variable called `@lexer` on the `Rouge::Lexers::Ruby` **singleton class**

Its clear now.. I got thrown off track when I read *"instance variable"*. Thanks.

**Comment by pyrmont** [comment]
I'm going to do more experimentation on my branch. Will submit a fresh PR if I make significant progress.
